crates/l1tx/src/filter/visitor.rs

```diff
@@ -0,0 +1,119 @@
+use bitcoin::{Block, Transaction};
+use strata_state::{
+    batch::SignedBatchCheckpoint,
+    tx::{DepositInfo, DepositRequestInfo, ProtocolOperation},
+};
+
+use super::{
+    parse_checkpoint_envelopes, parse_da_blobs, parse_deposit_requests, parse_deposits,
+    TxFilterConfig,
+};
+use crate::messages::ProtocolTxEntry;
+
+pub trait OpsVisitor {
+    // Do stuffs with `SignedBatchCheckpoint`.
+    fn visit_checkpoint(&mut self, _chkpt: SignedBatchCheckpoint) {}
+
+    // Do stuffs with `DepositInfo`.
+    fn visit_deposit(&mut self, _d: DepositInfo) {}
+
+    // Do stuffs with `DepositRequest`.
+    fn visit_deposit_request(&mut self, _d: DepositRequestInfo) {}
+
+    // Do stuffs with DA.
+    fn visit_da<'a>(&mut self, _d: impl Iterator<Item = &'a [u8]>) {}
+
+    fn collect(self) -> Vec<ProtocolOperation>;
+}
+
+pub trait BlockIndexer {
+    type Output;
+
+    fn collect(self) -> Self::Output;
+
+    fn index_tx(&mut self, txidx: u32, tx: &Transaction, config: &TxFilterConfig);
```

Comment:
Can you document the design for how these traits work and are expected to be used?  What purpose is there for these being separate traits?  Why does one have a impl-defined return type for `collect` but the other bakes in `Vec<ProtocolOperation>`?  Could these be consolidated into one?

Similarly, why is the `TxFilterConfig` being passed as an arg instead of being a member field that the indexer/visitor has itself.  Maybe we could give it a constructor that takes the filter config instead of passing it each time.

Reply:
Oops, totally forgot about the docs.

> Why does one have a impl-defined return type for collect but the other bakes in `Vec<ProtocolOperation>`?

The idea is that being an `OpsVisitor`, it makes more sense to bake `ProtocolOperation` into it. Howerver, for block indexer, it could extract anything from a block and hence that has an associated type. But from your comment above, we want to process da in a separate pass, so this might actually change.


>  Maybe we could give it a constructor that takes the filter config instead of passing it each time.

We could do that and I had thought about it but that feels like enforcing structure of the trait instance. So I made this method to take required context.

Reply:
> We could do that and I had thought about it but that feels like enforcing structure of the trait instance. So I made this method to take required context.

Perhaps, but then it makes it appear to be sensible to pass *different* filter configs on successive invocations, which it's *very not*.

Reply:
I agree, but that applies to every function with arguments that runs on a loop or a series of functions invocations that need to be passed the same object. Having those arguments makes it explicit that the method is supposed to primarily use those to extract information.